xen-hvm: Allocate some memory region in a reserved area

When allocating the VRAM, QEMU allocates it immediately after RAM. This
mostly works fine as there is a gap between the end of RAM and the start
of the MMIO region.  However, for instance when QEMU has 2G and the MMIO
region is 2G, there is no gap and QEMU will allocate the VRAM at 2G which
causes breakage since a BAR may be mapped to the same location.

Force RAM block address for guest memory allocation to always point to
a reserved area in guest address space.

They are only two memory regions that allocated with guest memory at the
moment, the VRAM and some memory for the TPM. Allocate a reserved address for
those.
diff --git a/exec.c b/exec.c
index ffdb518535..af8a3bf120 100644
--- a/exec.c
+++ b/exec.c
@@ -2211,8 +2211,8 @@ static void ram_block_add(RAMBlock *new_block, Error **errp, bool shared)
 
     if (!new_block->host) {
         if (xen_enabled()) {
-            xen_ram_alloc(new_block->offset, new_block->max_length,
-                          new_block->mr, &err);
+            new_block->offset = xen_ram_alloc(new_block->offset, new_block->max_length,
+                                              new_block->mr, &err);
             if (err) {
                 error_propagate(errp, err);
                 qemu_mutex_unlock_ramlist();
diff --git a/hw/i386/xen/xen-hvm.c b/hw/i386/xen/xen-hvm.c
index ab0b4e7f29..66eec5a7fc 100644
--- a/hw/i386/xen/xen-hvm.c
+++ b/hw/i386/xen/xen-hvm.c
@@ -261,23 +261,33 @@ static void xen_ram_init(PCMachineState *pcms,
     }
 }
 
-void xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size, MemoryRegion *mr,
+ram_addr_t xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size, MemoryRegion *mr,
                    Error **errp)
 {
     unsigned long nr_pfn;
     xen_pfn_t *pfn_list;
     int i;
 
+    // Only vga.vram MemoryRegion is moved, any other region are still at
+    // their starting location, even after migration.
+    if (mr->name && !strcmp(mr->name, "tpm-crb-cmd")) {
+        ram_addr = TPM_RESERVED_ADDRESS;
+    }
+
     if (runstate_check(RUN_STATE_INMIGRATE)) {
         /* RAM already populated in Xen */
         fprintf(stderr, "%s: do not alloc "RAM_ADDR_FMT
                 " bytes of ram at "RAM_ADDR_FMT" when runstate is INMIGRATE\n",
                 __func__, size, ram_addr);
-        return;
+        return ram_addr;
     }
 
     if (mr == &ram_memory) {
-        return;
+        return ram_addr;
+    }
+
+    if (mr->name && !strcmp(mr->name, "vga.vram")) {
+        ram_addr = VRAM_RESERVED_ADDRESS;
     }
 
     trace_xen_ram_alloc(ram_addr, size);
@@ -295,6 +305,7 @@ void xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size, MemoryRegion *mr,
     }
 
     g_free(pfn_list);
+    return ram_addr;
 }
 
 static XenPhysmap *get_physmapping(hwaddr start_addr, ram_addr_t size)
diff --git a/include/hw/xen/xen.h b/include/hw/xen/xen.h
index 894a0415b4..41933ad651 100644
--- a/include/hw/xen/xen.h
+++ b/include/hw/xen/xen.h
@@ -39,8 +39,8 @@ void xenstore_store_pv_console_info(int i, struct Chardev *chr);
 
 void xen_hvm_init(PCMachineState *pcms, MemoryRegion **ram_memory);
 
-void xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size,
-                   struct MemoryRegion *mr, Error **errp);
+ram_addr_t xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size,
+                         struct MemoryRegion *mr, Error **errp);
 void xen_hvm_modified_memory(ram_addr_t start, ram_addr_t length);
 
 void xen_register_framebuffer(struct MemoryRegion *mr);
@@ -48,4 +48,7 @@ void xen_register_framebuffer(struct MemoryRegion *mr);
 void xen_unassigned_access(hwaddr addr, bool is_write, void *val,
                            unsigned size);
 
+#define VRAM_RESERVED_ADDRESS  0xff000000
+#define TPM_RESERVED_ADDRESS   0xfed46000
+
 #endif /* QEMU_HW_XEN_H */
diff --git a/stubs/xen-hvm.c b/stubs/xen-hvm.c
index b7d53b5e2f..b1b06cb24b 100644
--- a/stubs/xen-hvm.c
+++ b/stubs/xen-hvm.c
@@ -35,9 +35,10 @@ int xen_is_pirq_msi(uint32_t msi_data)
     return 0;
 }
 
-void xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size, MemoryRegion *mr,
+ram_addr_t xen_ram_alloc(ram_addr_t ram_addr, ram_addr_t size, MemoryRegion *mr,
                    Error **errp)
 {
+    return 0;
 }
 
 qemu_irq *xen_interrupt_controller_init(void)
